### PR TITLE
Implement the octonions

### LINKED ISSUE #32940: Implement octonions and split octonions
<div id="comment:0"></div>

CC:  @slel

Component: **algebra**

Keywords: **octonions**

_Issue created by migration from https://trac.sagemath.org/ticket/32940_

### PR DESCRIPTION
**Pull request by tscrim**
### :books: Description

The octonions are a nonassociative normed division algebra, one of only 4 over the reals, and used to create the 7 dimensional cross product. We implement a slight generalization with 3 parameters over an arbitrary commutative ring of characteristic not equal to 2, which allows us to also define the split-octonions as another special case.

Fixes #32940.

### :memo: Checklist

- [x] The title is concise, informative, and self-explanatory.
- [x] The description explains in detail what this PR is about.
- [x] I have linked a relevant issue or discussion.
- [x] I have created tests covering the changes.
- [x] I have updated the documentation accordingly.

### CODE REVIEW

**Comment by fchapoton** [comment]
They do not seem to appear in the documentation. For instance under the setting "non-associative algebras".

**Comment by tscrim** [comment]
I thought I had fixed that. Well, now it is done. (With a force push since I had to rebase it anyways...)

**Comment by fchapoton** [comment]
There are several typos "octonian".

**Comment by tscrim** [comment]
Good catch(es). Thank you.

File: src/sage/algebras/octonion_algebra.pyx [hunk 1]
```diff
@@ -0,0 +1,988 @@
...
+class OctonionAlgebra(UniqueRepresentation, Parent):
+    r"""
+    The octonion algebra.
+
+    Let `R` be a commutative ring of characteristic not equal to `2`. The
+    *octonion algebra* with parameters `a, b, c` is a non-associative
+    non-commutative unital 8-dimensional `R`-algebra that is a deformation
+    of the usual octonions, which are when `a = b = c = -1`. The octonions
+    were originally constructed by Graves and independently discovered by
+    Cayley (who due to first publishing them, they are sometimes called
```
Comment (mkoeppe) [inline src/sage/algebras/octonion_algebra.pyx]: "who due to ..., they are" does not parse

File: src/sage/algebras/octonion_algebra.pyx [hunk 2]
```diff
@@ -0,0 +1,988 @@
...
+        from sage.misc.misc import some_tuples
+        for x, y in some_tuples(S, 2, tester._max_runs):
+            tester.assertEqual((x * x) * y, x * (x * y))
+            tester.assertEqual(y * (x * x), (y * x) * x)
+
+    def _test_hurwitz(self, **options):
+        r"""
+        Test that ``self`` is an Hurwitz algebra.
+
+        An algebra `A` is *Hurwitz* if there exists a quadratic form `N`
```
Comment (mkoeppe) [inline src/sage/algebras/octonion_algebra.pyx]: nondegenerate? Otherwise this is vacuous

**Review by fchapoton — APPROVED** [review]
ok, good to go

**Comment by tscrim** [comment]
Thank you!

**Pull request merged**